app/code/Magento/InventoryCatalogSearch/Plugin/Search/FilterMapper/AdaptStockStatusFilterPlugin.php

```diff
@@ -0,0 +1,200 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalogSearch\Plugin\Search\FilterMapper;
+
+use InvalidArgumentException;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\DB\Select;
+use Magento\CatalogSearch\Model\Search\FilterMapper\StockStatusFilter;
+use Magento\Framework\Search\Adapter\Mysql\ConditionManager;
+use Magento\InventoryIndexer\Model\StockIndexTableNameResolverInterface;
+use Magento\InventorySalesApi\Api\Data\SalesChannelInterface;
+use Magento\InventorySalesApi\Api\StockResolverInterface;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Adapt stock status filter to multi stocks
+ */
+class AdaptStockStatusFilterPlugin
+{
+    /**
+     * @var ConditionManager
+     */
+    private $conditionManager;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var StockResolverInterface
+     */
+    private $stockResolver;
+
+    /**
+     * @var StockIndexTableNameResolverInterface
+     */
+    private $stockIndexTableNameResolver;
+
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @param ConditionManager $conditionManager
+     * @param StoreManagerInterface $storeManager
+     * @param StockResolverInterface $stockResolver
+     * @param StockIndexTableNameResolverInterface $stockIndexTableNameResolver
+     * @param ResourceConnection $resourceConnection
+     */
+    public function __construct(
+        ConditionManager $conditionManager,
+        StoreManagerInterface $storeManager,
+        StockResolverInterface $stockResolver,
+        StockIndexTableNameResolverInterface $stockIndexTableNameResolver,
+        ResourceConnection $resourceConnection
+    ) {
+        $this->conditionManager = $conditionManager;
+        $this->storeManager = $storeManager;
+        $this->stockResolver = $stockResolver;
+        $this->stockIndexTableNameResolver = $stockIndexTableNameResolver;
+        $this->resourceConnection = $resourceConnection;
+    }
+
+    /**
+     * @param StockStatusFilter $subject
+     * @param callable $proceed
+     * @param Select $select
+     * @param array $stockValues
+     * @param string $type
+     * @param bool $showOutOfStockFlag
+     * @return Select
+     * @throws \InvalidArgumentException
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function aroundApply(
+        StockStatusFilter $subject,
+        callable $proceed,
+        Select $select,
+        $stockValues,
+        $type,
+        $showOutOfStockFlag
+    ) {
+        if ($type !== StockStatusFilter::FILTER_JUST_ENTITY
+            && $type !== StockStatusFilter::FILTER_ENTITY_AND_SUB_PRODUCTS
+        ) {
+            throw new InvalidArgumentException('Invalid filter type: ' . $type);
+        }
+
+        $select = clone $select;
```

Comment:
no need to clone select especially taking into account that we don't call proceed method (method being pluginized)